src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java

```diff
@@ -0,0 +1,597 @@
+     *
+     * @param competency the competency for which the utility should be computed
+     * @param state      the current state of the recommendation system
+     * @return extends or assumes utility of the competency
+     */
+    private static double computeExtendsOrAssumesUtility(Competency competency, RecommendationState state) {
+        final double weight = state.extendsCompetencies.get(competency.getId()) * EXTENDS_UTILITY_RATIO + state.assumesCompetencies.get(competency.getId()) * ASSUMES_UTILITY_RATIO;
+        // return max utility if competency does not extend or assume other competencies
+        if (weight == 0) {
+            return EXTENDS_OR_ASSUMES_UTILITY;
+        }
+        return (1 / weight) * EXTENDS_OR_ASSUMES_UTILITY;
+
+    }
+
+    /**
+     * Gets the utility of the competency with respect to users mastery progress within the competency.
+     *
+     * @param competency the competency for which the utility should be computed
+     * @param state      the current state of the recommendation system
+     * @return mastery utility of the competency
+     */
+    private static double computeMasteryUtility(Competency competency, RecommendationState state) {
+        return state.competencyMastery.get(competency.getId()) * MASTERY_PROGRESS_UTILITY;
+    }
+
+    /**
+     * Analyzes the current progress within the learning path and generates a recommended ordering of learning objects in a competency.
+     *
+     * @param learningPath the learning path that should be analyzed
+     * @param competency   the competency
+     * @param state        the current state of the recommendation
+     * @return the recommended ordering of learning objects
+     */
+    public List<LearningObject> getRecommendedOrderOfLearningObjects(LearningPath learningPath, Competency competency, RecommendationState state) {
+        var pendingLectureUnits = competency.getLectureUnits().stream().filter(lectureUnit -> !lectureUnit.isCompletedFor(learningPath.getUser())).toList();
+        ArrayList<LearningObject> recommendedOrder = new ArrayList<>(pendingLectureUnits);
+
+        // early return if competency can be trivially mastered
+        if (canBeMasteredWithoutExercises(competency)) {
+            return recommendedOrder;
+        }
+
+        final var combinedPriorConfidence = computeCombinedPriorConfidence(competency, state);
+        final var pendingExercises = competency.getExercises().stream().filter(exercise -> !learningObjectService.isCompletedByUser(exercise, learningPath.getUser()))
+                .collect(Collectors.toSet());
+        final var numberOfExercisesRequiredToMaster = predictNumberOfExercisesRequiredToMaster(learningPath, competency, combinedPriorConfidence, pendingExercises.size());
+        log.warn("numberOfExercisesRequiredToMaster:" + numberOfExercisesRequiredToMaster);
+        HashMap<DifficultyLevel, Set<Exercise>> difficultyLevelMap = generateDifficultyLevelMap(competency.getExercises());
```

Comment:
```suggestion
        Map<DifficultyLevel, Set<Exercise>> difficultyLevelMap = generateDifficultyLevelMap(competency.getExercises());
```
Same as for the list above